File: game-of-life/typescript/src/Grid.ts
import { type Cell, cellKey, parseKey, neighbors } from './Cell.js';

/**
 * Immutable set of living cells on an unbounded infinite plane.
 * `tick()` applies the four GoL rules and returns the next generation.
 */
export class Grid {
  private readonly cells: ReadonlySet<string>;

  constructor(livingCells: Iterable<Cell>) {
    const keys = new Set<string>();
    for (const cell of livingCells) {
      keys.add(cellKey(cell));
    }
    this.cells = keys;
  }

  isAlive(row: number, col: number): boolean {
    return this.cells.has(cellKey({ row, col }));
  }

  livingCells(): Cell[] {
    return [...this.cells].map(parseKey);
  }

  tick(): Grid {
    const neighborCounts = new Map<string, number>();

    for (const key of this.cells) {
      const cell = parseKey(key);
      for (const neighbor of neighbors(cell)) {
        const nk = cellKey(neighbor);
        neighborCounts.set(nk, (neighborCounts.get(nk) ?? 0) + 1);
      }
    }

    const nextGeneration: Cell[] = [];

    for (const [key, count] of neighborCounts) {
      const isAlive = this.cells.has(key);
      if (count === 3 || (count === 2 && isAlive)) {
        nextGeneration.push(parseKey(key));
      }
    }

    return new Grid(nextGeneration);
  }
}

File: game-of-life/csharp/src/GameOfLife/Grid.cs
namespace GameOfLife;

/// <summary>
/// Immutable set of living cells on an unbounded infinite plane.
/// <c>Tick()</c> applies the four GoL rules and returns the next generation.
/// </summary>
public class Grid
{
    private readonly HashSet<Cell> _livingCells;

    public Grid(IEnumerable<Cell> livingCells)
    {
        _livingCells = new HashSet<Cell>(livingCells);
    }

    public bool IsAlive(int row, int col) => _livingCells.Contains(new Cell(row, col));

    public IReadOnlySet<Cell> LivingCells => _livingCells;

    public Grid Tick()
    {
        // Count neighbors for every cell adjacent to a living cell.
        var neighborCounts = new Dictionary<Cell, int>();

        foreach (var cell in _livingCells)
        {
            foreach (var neighbor in cell.Neighbors())
            {
                neighborCounts.TryGetValue(neighbor, out var count);
                neighborCounts[neighbor] = count + 1;
            }
        }

        var nextGeneration = new HashSet<Cell>();

        foreach (var (candidate, count) in neighborCounts)
        {
            var isAlive = _livingCells.Contains(candidate);

            // Survival: live cell with 2 or 3 neighbors.
            // Reproduction: dead cell with exactly 3 neighbors.
            if (count == 3 || (count == 2 && isAlive))
            {
                nextGeneration.Add(candidate);
            }
        }

        return new Grid(nextGeneration);
    }
}
